README.md
# Cisco Configuration Manager #

This project provides tools to manage configurations for Cisco devices. 
At the heart of the project is a tftp daemon written entirely in PHP.
Configurations are read and written to the tftp server using a file path
that incorporates a password feature for security. The current version of
each device configuration is stored in a mysql database. Subsequent changes
to the configurations are stored in diff format to provide revision history.
A web frontend is provided to manage devices and view configurations and
history.

The docs/README.txt file is essential reading. It outlines the system
requirements including required PHP5 compile options and PECL extensions.

The docs/server_build.txt outlines a server build for cisco-config-manager
from scratch. It assumes a VMware environment although this is by no means
a requirement. The accompanying docs/php_build.sh script is a quick way
to get PHP5 configuration that supports cisco-config-manager.

This project was conceived to address some specific needs at K-Net
<http://www.knet.ca/> where it is used on a daily basis. Be warned, the
documentation and installation instructions are sparse.

This work is released under the GNU General Public License, please consult
LICENSE and <http://www.gnu.org/licenses/old-licenses/gpl-2.0.html>.



CHANGES
# CHangelog #

Nov 12, 2014
o Now supports php-5.3.29
o Applied minor fixes to bring up to php-5.3.29
o Added server build instruction
o Added php build script



db/db.php
<?php
/* $Id: db.php,v 1.3 2005-11-18 23:28:56 adicvs Exp $
 * 
 * Copyright (C) 2005 Adi Linden <adi@adis.ca>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

/* db.php   A simple class to access data in a mysql database. It provides
 *          some simple functions that process query results in easier to
 *          handle formats (associative arrays) then native php mysql 
 *          functions.
 *
 *          This class inspired and based on ezSQL by Justin Vincent. For
 *          a more featureful solution see ezSQL.
 */

class db {

    /* Class constructor
     *
     * Opens database connection and selects the database we want to work
     * with.
     */
    function __construct ($host, $user, $pass, $dbname)
    {
        if (! $this->open($host, $user, $pass)) {
            return false;
        }
        if (! $this->select($dbname)) {
            return false;
        }
        return true;
    }

    /* Class destructor
     *
     * Closes the database connection.
     */
    function __destruct ()
    {
        $this->close();
    }

    /* The database query
     *
     * This sends $query to the database connection. Returns the number 
     * of rows affected by the operation. The $this->result object can be 
     * used to obtain results outside this class.
     */
    function query ($query)
    {
        /* Run query */
        $this->result = @mysql_query($query, $this->dbh);
        if (! $this->result) {
            echo "Error sending query: ".mysql_error();
            return false;
        }

        /* Determine number of rows affected */
        if ( preg_match("/^(insert|delete|update|replace)\s+/i",$query) ) {
            $n = mysql_affected_rows($this->dbh);
        } else {
            $n = mysql_num_rows($this->result);
        }
        return $n;
    }

    /* Get a single row
     *
     * Sends the (optional) $query to the database connection. If $query is 
     * not specified it operates on $this->result of a previous operation.
     * Returns an associative array containing the result of the first row. 
     */
    function get_row ($query = NULL)
    {
        /* Send the query to the db */
        if ($query) {
            $this->query($query);
        }
        
        /* Process the result */
        $r = mysql_fetch_assoc($this->result);

        return $r;
    }

    /* Get multiple rows
     *
     * Sends the (optional) $query to the database connection. If $query is 
     * not specified it operates on $this->result of a previous operation.
     * Returns an multi dimensional array containing the resulting rows as 
     * numeric indices and fields as associative nested array.
     */
    function get_rows ($query = NULL)
    {
        /* Send the query to the db */
        if ($query) {
            $this->query($query);
        }
        
        /* Process the result */
        $i = 0;
        while ($r = mysql_fetch_assoc($this->result)) {
            $a[$i] = $r;
            $i++;
        }
        return $a;
    }

    /* Open database connection */
    function open ($host, $user, $pass)
    {
        $this->dbh = @mysql_connect($host, $user, $pass);
        if (! $this->dbh) {
            echo "Error connecting to db: ".mysql_error();
            return false;
        }
        return true;
    }

    /* Select database */
    function select($dbname)
    {
        if ( !@mysql_select_db($dbname, $this->dbh)) {
            echo "Error selecting db: ".$dbname;
            return false;
        }
        return true;
    }

    /* Close database connection */
    function close ()
    {
        @mysql_close($this->dbh);
    }

}
?>


browse.php
<?php
/* Copyright (C) 2005-2014 Adi Linden <adi@adis.ca>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

/* browse.php   Browse the /tftpboot directory structure.
 */

/* Include configuration */
require_once('./config.php');

/* Get path */
if (isset($_GET['path'])) {
    $dir = sanitize($_GET['path']);
} else {
    $dir = TFTP_FILE_ROOT;
}

/* Download or show files */
if (isset($_GET['download'])) {
    /* Download files */
    $file = sanitize($_GET['download']);
    get_file($file);
} else {
    /* Show files */
    send_header($dir);
    show_files($dir);
    send_footer();
}

function sanitize($path)
{
    /* Don't get back up the directory structure */
    if (preg_match('/\.\./', $path) > 0) {
        echo "Illegal path specified";
        exit;
    }
    /* Make sure we are within DIRROOT */
    if (preg_match('!^'.TFTP_FILE_ROOT.'!', $path) == 0) {
        echo "Illegal access";
        exit;
    }
    return $path;
}

function get_file($file)
{
    $fp = @fopen($file, 'r');
    if (! $fp) {
        echo "Permission denied";
        return false;
    }

    /* Download all files except .txt */
    header("Expires: Mon, 26 Jul 1997 05:00:00 GMT"); // some day in the past
    header("Last-Modified: " . gmdate("D, d M Y H:i:s") . " GMT");
    if (preg_match('/\.txt$/', $file) > 0) {
        header("Content-type: text/plain");
    } else {
        header("Content-type: application/x-download");
        header("Content-Disposition: attachment; filename=" . basename($file));
        header("Content-Transfer-Encoding: binary");
    }
    while (! feof($fp)) {
        print(fread($fp,1024));
        flush();
    }
    fclose($fp);
}

function do_ls($dir) {
    $dh = @opendir($dir);
    if ($dh) {
        while(false !== ($files[] = readdir($dh))) {
        }
        return $files;
    }
    return false;
}

function show_files($dir)
{
    if (! $files = do_ls($dir)) {
        echo "Failed to access location\n";
        return false;
    }

    asort($files);
    foreach ($files as $file) {
        /* Ignore dots and empti files */
        if ($file == '' || $file == '.' || $file == '..') {
            continue;
        }
        /* Prepend dir */
        $path = $dir . '/' . $file;
        /* Show file */
        if (is_file($path)) {
            echo '<img src="./icons/generic.png" alt="file">&nbsp;&nbsp;';
            echo '<a href="'.$_SERVER['PHP_SELF'].'?download='.$path.'">'.$file."</a><br>\n";
        }
        /* Show dir */
        if (is_dir($path)) {
            echo '<img src="./icons/dir.png" alt="dir">&nbsp;&nbsp;';
            echo '<a href="'.$_SERVER['PHP_SELF'].'?path='.$path.'">'.$file."</a><br>\n";
        }
    }
}

function send_header($dir)
{
?><!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
<head>
  <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
  <title>Browse: <?php echo $dir; ?></title>
</head>
<body>
<h2>Browse: <?php echo $dir; ?></h2>
<p>
<?php 
}   
      
function send_footer()
{   
?>  
</body>
</html>
<?php 
}     

?>


config.php
<?php
/* Copyright (C) 2014 Adi Linden <adi@adis.ca>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

/* devices.php  Manage devices, view configurations and configuration
 *              history.
 */

/* Database credentials */
define ('TFTP_DB',              'ccm');
define ('TFTP_DB_HOST',         'localhost');
define ('TFTP_DB_USER',         'ccm_user');
define ('TFTP_DB_PASS',         'schmack');

/* File access */
define('TFTP_FILE_ROOT',        '/tftpboot');
define('TFTP_FILE_RO',          '1');

/* Delete option */
define('DELETE_ON_LIST',        '0');
define('DELETE_ON_EDIT',        '1');
?>


tftpd/tftpd-file.php
<?php
/* Copyright (C) 2005-2014 Adi Linden <adi@adis.ca>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

/* tftpd-file.php   This file is included by tftpd.php. I contains the 
 *                  functions that are specific to accessing the local 
 *                  file system. 
 */

/* 
 *  Do NOT run this script through a browser. 
 *  Needs to be accessed from a shell.
 */
if ($_SERVER["SHELL"] != '/bin/bash' && $_SERVER["SHELL"] != '/bin/sh') {
    die("<br><strong>This script is cannot be run through browser!</strong>");
}

/* Include configuration */
require_once('../config.php');

/* Handle file requests */
function tftpd_handle_file ($c_sock, $sock, $opcode, $mode, $path, $file)
{
    /* Accessing file in TFTP_FILE_ROOT */
    if (!preg_match('/\/$/', $path) && $path != '') {
        $path .= '/';
    }
    $file = TFTP_FILE_ROOT . '/' . $path . $file;
    tftpd_log('1', 'access file: '.$file);

    /* Deny any accesses to paths containing ../ */
    if (preg_match('/\.\.\//', $file)) {
        tftpd_send_nak($c_sock, $sock, TFTP_EACCESS, 'path contains ../');
        return false;
    }
    
    /* Check general file permissions, we can't stat() a non existing file
     * See stat.h for details on the stat() bit masks.
     *
     * Need to use lstat() to check symlink, stat() to check file the
     * symlink points to. !!! To allow symlinks change to sast() !!!
     */
    if (! $stat = @lstat($file)) {
        tftpd_send_nak($c_sock, $sock, TFTP_ENOTFOUND, 'cannot stat');
        return false;
    }

    /* Don't allow symlinks 
     * #define S_IFMT  00170000
     * #define S_IFLNK  0120000
     * #define S_ISLNK(m)      (((m) & S_IFMT) == S_IFLNK)
     */
    if (($stat['mode'] & 00170000) === 0120000) {
        tftpd_send_nak($c_sock, $sock, TFTP_EACCESS, 'file is symlink');
        return false;
    }
    /* Handle read request */
    if ($opcode == TFTP_RRQ) {
        /* Is file publically readable?
         * #define S_IROTH 00004
         */
        if (($stat['mode'] & 00004) === 00 ) {
            tftpd_send_nak($c_sock,$sock,TFTP_EACCESS,'no public read access');
            return false;
        }
        /* Open file for reading */
        $fp = fopen($file, 'r', false);
        if (! $fp) {
            tftpd_send_nak($c_sock, $sock, TFTP_UNDEF, 'fopen() failed');
            return false;
        }
        /* Send file to peer */
        tftpd_send_file($c_sock, $sock, $fp);
        fclose($fp);
        return true;
    }

    /* Handle write request */
    if ($opcode == TFTP_WRQ) {
        /* Is file publically writable?
         * #define S_IWOTH 00002
         */
        if (($stat['mode'] & 00002) === 00) {
            tftpd_send_nak($c_sock,$sock,TFTP_EACCESS,'no public write access');
            return false;
        }
        /* Check read-only flag */
        if (TFTP_FILE_RO) {
            tftpd_send_nak($c_sock,$sock,TFTP_EACCESS,'write access disabled');
            return false;
        }
        /* Open file for writing */
        $fp = fopen($file, 'w', false);
        if (! $fp) {
            tftpd_send_nak($c_sock, $sock, TFTP_UNDEF,'fopen() failed');
            return false;
        }
        /* Receive file from peer */
        tftpd_recv_file($c_sock, $sock, $fp);
        fclose($fp);
        return true;
    }

    /* We should never get here */
    fclose($fp);
    tftpd_send_nak($c_sock, $sock, TFTP_EBADOP, 'ouch! not good...');
    return false;
}

/* Receive file */
function tftpd_recv_file($s, $sock, $fp)
{
    $block = 0;
    $xfer_byte = 0;
    $xfer_time = tftpd_microtime();
    do {
        if (! tftpd_receive_data($s, $sock, $block, $data)) {
            return false;
        }
        $r = fwrite($fp, $data);
        if ($r != strlen($data)) {
            tftpd_send_nak($c_sock, $sock, TFTP_ENOSPACE, 'disk full?');
            return false;
        }
        $block++;
        if ($block > 65535) {
            $block = 0;
        }
        $xfer_byte += strlen($data);
    } while (strlen($data) == 512);

    /* Be a good citizen and churn out one last ACK */
    $s_buf = pack('nn', TFTP_ACK, $block);
    tftpd_send_packet($s, $sock, $s_buf);

    /* Log our success */
    $xfer_time = round(tftpd_microtime() - $xfer_time, 3);
    tftpd_log('1', 'received '.$xfer_byte.' bytes in '.$xfer_time.' seconds');
}

/* Send file */
function tftpd_send_file($s, $sock, $fp)
{
    $block = 1;
    $xfer_byte = 0;
    $xfer_time = tftpd_microtime();
    while (!feof($fp)) {
        $data = fread($fp, TFTP_SEGSIZE);
        if (! tftpd_send_data($s, $sock, $block, $data)) {
            return false;
        }
        $block++;
        if ($block > 65535) {
            $block = 0;
        }
        $xfer_byte += strlen($data);
    }

    /* Log our success */
    $xfer_time = round(tftpd_microtime() - $xfer_time,3);
    tftpd_log('1', 'sent '.$xfer_byte.' bytes in '.$xfer_time.' seconds');
}

?>


devices.php
<?php
/* Copyright (C) 2005-2014 Adi Linden <adi@adis.ca>
 *
 * This program is free software; you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation; either version 2 of the License, or
 * (at your option) any later version.
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program; if not, write to the Free Software
 * Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
 */

/* devices.php  Manage devices, view configurations and configuration
 *              history.
 */

/* Include configuration */
require_once('./config.php');

/* Include database access library */
require_once('./db/db.php');

/* Open database connextion */
$db = new db(TFTP_DB_HOST, TFTP_DB_USER, TFTP_DB_PASS, TFTP_DB);

/* Process form vars */
$vars = array('action','id','devicename','description','filename','password','ts','patch');
foreach($vars as $var) {
    $vv = "db_".$var;
    if (isset($_POST[$var])) {
        $$var = trim($_POST[$var]);
        $$vv = mysql_real_escape_string($$var);
    } else if (isset($_GET[$var])) {
        $$var = trim($_GET[$var]);
        $$vv = mysql_real_escape_string($$var);
    }
    else {
        $$var = '';
        $$vv = '';
    }
}

/* Process request */
switch ($action) {

    /* Single patch */
    case 'patch':
        if ($id == '') {
            $message[] = "No device specified. ";
            devices_header(NULL, $message);
            devices_table();
            devices_footer();
            break;
        }
        devices_header('View Patch');
        devices_view_patch($db_id, $db_patch);
        devices_footer();
        break;

    /* Single historic configuration */
    case 'history':
        if ($id == '') {
            $message[] = "No device specified. ";
            devices_header(NULL, $message);
            devices_table();
            devices_footer();
            break;
        }
        devices_header('View Historic Configuration');
        devices_view_history($db_id, $db_ts);
        devices_footer();
        break;

    /* List of patches for device */
    case 'patches':
        if ($id == '') {
            $message[] = "No device specified. ";
            devices_header(NULL, $message);
            devices_table();
            devices_footer();
            break;
        }
        devices_header('View Device History');
        devices_list_patches($db_id);
        devices_footer();
        break;

    /* Current configuration */
    case 'view':
        if ($id == '') {
            $message[] = "No device specified. ";
            devices_header(NULL, $message);
            devices_table();
            devices_footer();
            break;
        }
        devices_header('View Device Configuration');
        devices_view($db_id);
        devices_footer();
        break;

    /* Add device form */
    case 'add':
        devices_header('Add New Device');
        devices_edit();
        devices_footer();
        break;

    /* Add device */
    case 'added':
        unset($message);
        if ($devicename == '') {
            $message[] = "Device name is required. ";
        }
        if ($filename == '') {
            $message[] = "File name is required. ";
        }
        $n = $db->query("SELECT id FROM devices WHERE devicename='$db_devicename'");
        if ($n > 0) {
            $message[] = "Duplicate device name '$devicename' not allowed. ";
        }
        $n = $db->query("SELECT id FROM devices WHERE filename='$db_filename'");
        if ($n > 0) {
            $message[] = "Duplicate file name '$filename' not allowed. ";
        }
        
        if (! isset($message)) {
            /* INSERT new device */
            $db->query("INSERT INTO devices (devicename,description,filename,password) 
                    VALUES ('$db_devicename','$db_description','$db_filename','$db_password')");
            $message[] = "Device added. ";
            devices_header('List Devices',$message);
            devices_table();
        } else {
            devices_header('Add New Device',$message);
            devices_edit();
        }
        devices_footer();
        break;

    /* Edit device form */
    case 'edit':
        devices_header('Edit Device');
        devices_edit($id);
        devices_footer();
        break;

    /* Edit device */
    case 'edited':
        unset($message);
        if ($id == '') {
            $message[] = "Bandits! This ain't supposed to be! ";
        }
        if ($devicename == '') {
            $message[] = "Device name is required. ";
        }
        if ($filename == '') {
            $message[] = "File name is required. ";
        }
        $r = $db->get_row("SELECT id FROM devices WHERE devicename='$db_devicename'");
        if ($r['id'] != '' && $r['id'] != $id) {
            $message[] = "Duplicate device name '$devicename' not allowed. ";
        }
        $r = $db->get_row("SELECT id FROM devices WHERE filename='$db_filename'");
        if ($r['id'] != '' && $r['id'] != $id) {
            $message[] = "Duplicate file name '$filename' not allowed. ";
        }

        if (! isset($message)) {
            /* Update database */
            $db->query("UPDATE devices SET devicename='$db_devicename',description='$db_description',
                    filename='$db_filename',password='$db_password' WHERE id='$db_id'");
            $message[] = "Device updated. ";
            devices_header('List Devices',$message);
            devices_table();
        } else {
            devices_header('Edit Device',$message);
            devices_edit();
        }

        devices_footer();
        break;

    /* Delete device */
    case 'delete':
        unset($message);
        if ($db_id == '') {
            $message[] = "Bandits! This ain't supposed to be! ";
        }
        $n = $db->query("SELECT devicename FROM devices WHERE id='$db_id'");
        if ($n < 1) {
            $message[] = "Device not found! ";
        }
        if (! isset($message)) {
            /* Device information */
            $r = $db->get_row("SELECT devicename,description,filename,password FROM devices WHERE id='$id'");
            /* Delete from tables */
            $db->query("DELETE FROM address WHERE device='$db_id'");
            $db->query("DELETE FROM configurations WHERE device='$db_id'");
            $db->query("DELETE FROM devices WHERE id='$db_id'");
            $db->query("DELETE FROM patches WHERE device='$db_id'");
            $db->query("DELETE FROM updated WHERE device='$db_id'");
            $message[] = "Deleted ".$r['devicename']."! ";
        }
        devices_header('List Devices', $message);
        devices_table();
        devices_footer();
        break;

    /* List devices */
    default:
        devices_header();
        devices_table();
        devices_footer();
        break;
}

unset($db);
/* End of main script, just functions below */

function devices_view_patch($id, $patch)
{
    global $db;

    /* Device information */
    $r = $db->get_row("SELECT devicename,description,filename,password FROM devices WHERE id='$id'");
?>
<p><a href="<?php echo $_SERVER['PHP_SELF']; ?>">&laquo; List Devices</a>
<a href="<?php echo $_SERVER['PHP_SELF']; ?>?action=view&id=<?php echo $id; ?>">&laquo; Configuration</a>
<a href="<?php echo $_SERVER['PHP_SELF']; ?>?action=patches&id=<?php echo $id; ?>">&laquo; History</a>
<p><strong>Device Name:</strong>  <?php echo $r['devicename']; ?>
<br><strong>File Name:</strong>  <?php echo $r['filename']; ?>
<p><strong>Patch:</strong>
<?php
    /* Get revision for the patch by walking the array */
    $rows = $db->get_rows("SELECT id FROM patches WHERE device='$id' ORDER BY ts ASC");
    for ($rev = count($rows) - 1; $rev >= 0; $rev--) {
        $row = $rows[$rev];
        /* Array with patch id as key and revision as value */
        $patches[$row['id']] = $rev;
    }

    /* Retrieve patch */
    $row = $db->get_row("SELECT ts,content FROM patches WHERE id='$patch'");
    if ($row['ts'] == '') {
        echo "<p><strong>No matching revision found!</strong>\n";
    } else {
        echo "<br>revision: ".$patches[$patch].", date: ".devices_ts_to_date($row['ts'],'%Y-%m-%d %H:%i:%s').", ";
        echo "<pre>\n".$row['content']."</pre>\n";
    }
?>
<p><a href="<?php echo $_SERVER['PHP_SELF']; ?>">&laquo; List Devices</a>
<a href="<?php echo $_SERVER['PHP_SELF']; ?>?action=view&id=<?php echo $id; ?>">&laquo; Configuration</a>
<a href="<?php echo $_SERVER['PHP_SELF']; ?>?action=patches&id=<?php echo $id; ?>">&laquo; History</a>
<?php
}

function devices_view_history($id, $ts)
{
    global $db;

    /* Device information */
    $r = $db->get_row("SELECT devicename,description,filename,password FROM devices WHERE id='$id'");
?>
<p><a href="<?php echo $_SERVER['PHP_SELF']; ?>">&laquo; List Devices</a>
<a href="<?php echo $_SERVER['PHP_SELF']; ?>?action=view&id=<?php echo $id; ?>">&laquo; Configuration</a>
<a href="<?php echo $_SERVER['PHP_SELF']; ?>?action=patches&id=<?php echo $id; ?>">&laquo; History</a>
<p><strong>Device Name:</strong>  <?php echo $r['devicename']; ?>
<br><strong>File Name:</strong>  <?php echo $r['filename']; ?>
<p><strong>Historic Configuration:</strong>
<p>
<?php
    /* Get current config */
    $r = $db->get_row("SELECT content FROM configurations WHERE device='$id'");
    $config = $r['content'];
    echo "-&gt; retrieving current configuration<br>\n";

    /* Apply patches */
    $rows = $db->get_rows("SELECT ts,content FROM patches WHERE device='$id' ORDER BY ts ASC");
    $rev = count($rows) - 1;
    for ($rev; $rev >= 0; $rev--) {
        $row = $rows[$rev];
        /* Only apply patches that are newer then the requested date */
        if ($row['ts'] >= $ts) {
            echo "-&gt; applying patch revision ".$rev.", date ".devices_ts_to_date($row['ts'])."<br>\n";
            $config = xdiff_string_patch($config, $row['content']);
        }
    }
?>
<pre>
<?php echo $config; ?>
</pre>
<p><a href="<?php echo $_SERVER['PHP_SELF']; ?>">&laquo; List Devices</a>
<a href="<?php echo $_SERVER['PHP_SELF']; ?>?action=view&id=<?php echo $id; ?>">&laquo; Configuration</a>
<a href="<?php echo $_SERVER['PHP_SELF']; ?>?action=patches&id=<?php echo $id; ?>">&laquo; History</a>
<?php 
}

function devices_list_patches($id)
{
    global $db;

    $r = $db->get_row("SELECT devicename,description,filename,password FROM devices WHERE id='$id'");
?>
<p><a href="<?php echo $_SERVER['PHP_SELF']; ?>">&laquo; List Devices</a>
<a href="<?php echo $_SERVER['PHP_SELF']; ?>?action=view&id=<?php echo $id; ?>">&laquo; Configuration</a>
<p><strong>Device Name:</strong>  <?php echo $r['devicename']; ?>
<br><strong>Description:</strong>  <?php echo $r['description']; ?>
<br><strong>File Name:</strong>  <?php echo $r['filename']; ?>
<p><strong>History:</strong>
<?php
    $rows = $db->get_rows("SELECT id,ts FROM patches WHERE device='$id' ORDER BY ts ASC");
    $rev = count($rows);
    /* Current config */
    $view_config = '<a href="'.$_SERVER['PHP_SELF'].'?action=history&ts=20991122002647&id='.$id.'">view</a>';
    echo "<p>revision: ".$rev.", ";
    echo $view_config." current\n";
    $rev--;
    /* Patches */
    for ($rev; $rev >= 0; $rev--) {
        $row = $rows[$rev];
        $view_config = '<a href="'.$_SERVER['PHP_SELF'].'?action=history&ts='.$row['ts'].'&id='.$id.'">view</a>';
        $view_patch = '<a href="'.$_SERVER['PHP_SELF'].'?action=patch&patch='.$row['id'].'&id='.$id.'">view</a>';
        echo "<p>revision: ".$rev.", date: ".devices_ts_to_date($row['ts'],'%Y-%m-%d %H:%i:%s').", ";
        echo $view_config." configuration";
        echo ", ";
        echo $view_patch." patch";
        echo "\n";
    }
?>
<p><a href="<?php echo $_SERVER['PHP_SELF']; ?>">&laquo; List Devices</a>
<a href="<?php echo $_SERVER['PHP_SELF']; ?>?action=view&id=<?php echo $id; ?>">&laquo; Configuration</a>
<?php
}

function devices_view($id)
{
    global $db;

    $r = $db->get_row("SELECT devicename,description,filename,password FROM devices WHERE id='$id'");
?>
<p><a href="<?php echo $_SERVER['PHP_SELF']; ?>">&laquo; List Devices</a>
<p><strong>Device Name:</strong>  <?php echo $r['devicename']; ?>
<br><strong>Description:</strong>  <?php echo $r['description']; ?>
<br><strong>File Name:</strong>  <?php echo $r['filename']; ?>
<br><strong>Password:</strong>  <?php echo $r['password']; ?>
<p><strong>Cisco IOS command to save config:</strong>
<pre>copy running-config tftp://<?php echo $_SERVER['SERVER_ADDR'] ?>/$<?php echo $r['password']; ?>$/<?php echo $r['filename']; ?></pre>
<p><strong>Older Cisco IOS configuration:</strong>
<pre>alias exec save copy system:/running-config tftp://<?php echo $_SERVER['SERVER_ADDR'] ?>/$<?php echo $r['password']; ?>$/<?php echo $r['filename']; ?></pre>
<p><strong>Recommended Cisco IOS configuration:</strong>
<pre>alias exec save show running-config | redirect tftp://<?php echo $_SERVER['SERVER_ADDR'] ?>/$<?php echo $r['password']; ?>$/<?php echo $r['filename']; ?></pre>
<?php
    $r = $db->get_row("SELECT ts FROM updated WHERE device='$id'");
?>
<p><strong>Configuration:</strong>
<br>Saved on <?php echo devices_ts_to_date($r['ts']); ?>
<?php
    $r = $db->query("SELECT ts FROM patches WHERE device='$id'");
    if ($r > 0) :
?>
 (<a href="<?php echo $_SERVER['PHP_SELF'].'?action=patches&id='.$id; ?>">View history</a>)
<?php
    endif;
?>
<pre>
<?php 
    $r = $db->get_row("SELECT content FROM configurations WHERE device='$id'");
    echo htmlentities($r['content'], ENT_QUOTES); 
?>
</pre>
<p><a href="<?php echo $_SERVER['PHP_SELF']; ?>">&laquo; List Devices</a>
<?php
}

function devices_edit($id = NULL)
{
    global $db;

    if (! $id) {
        $action = 'added';
        $device['password'] = devices_gen_pwd();
    } else {
        $action = 'edited';
        $device = $db->get_row("SELECT devicename,description,filename,password FROM devices WHERE id='".$id."'");
    }

    ?>
<p>
<form action="<?php echo $_SERVER['PHP_SELF']; ?>" method="post">
  <p>Device Name:<br><input type="text" size="60" name="devicename" value="<?php if (isset($device['devicename'])) { echo $device['devicename']; }  ?>">
  <p>Description:<br><input type="text" size="60" name="description" value="<?php if (isset($device['description'])) { echo $device['description']; } ?>">
  <p>File Name:<br><input type="text" size="60" name="filename" value="<?php if (isset($device['filename'])) { echo $device['filename']; } ?>">
  <p>Password:<br><input type="text" size="60" name="password" value="<?php echo $device['password'] ?>">
  <p><input type="hidden" name="action" value="<?php echo $action ?>">
    <input type="hidden" name="id" value="<?php echo $id ?>">
    <?php
    echo '<a href="'.$_SERVER['PHP_SELF'].'"><button type="button">&laquo; Back</button></a>';
    echo '&nbsp;--&nbsp;';
    if (DELETE_ON_EDIT && $id) {
        echo '<a onclick="return confirm(\'Are you sure you want to delete device: '.$device['devicename'].'?\')" '.
             'href="'.$_SERVER['PHP_SELF'].'?action=delete&id='.$id.'">'.
             '<button type="button">Delete</button>'.
             '</a>';
        echo '&nbsp;--&nbsp;';
    }
    echo '<input type="submit" value="Update &raquo;">'."\n";
    ?>
</form>
    <?php
}

function devices_table()
{
    global $db;

    $rows = $db->get_rows("SELECT id,devicename,description FROM devices ORDER BY devicename ASC");
?>
<p><a href="<?php echo $_SERVER['PHP_SELF']; ?>?action=add">Add</a> a new device.
<p><table width="70%" border="0" cellpadding="2" cellspacing="2">
<tr>
  <th>Device</th>
  <th>Description</th>
<?php
    if (DELETE_ON_LIST) {
        echo '  <th colspan="3">Action</th>';
    } else {
        echo '  <th colspan="2">Action</th>';
    }
?>
</tr>
<?php

    /* Process each device */
    $alt = '';
    foreach ($rows as $row) {
        if ($alt == '') {
            $alt = 'alt';
        } else {
            $alt = '';
        }
        echo "<tr class='".$alt."'>\n  <td>".$row['devicename']."</td>\n";
        echo "  <td>".$row['description']."</td>\n";
        echo "  <td class='edit'>".
            '<a href="'.$_SERVER['PHP_SELF'].'?action=edit&id='.$row['id'].'" class="edit">Edit</a>'.
            "</td>\n";
        echo "  <td class='view'>".
            '<a href="'.$_SERVER['PHP_SELF'].'?action=view&id='.$row['id'].'" class="view">View</a>'.
            "</td>\n";
        if (DELETE_ON_LIST) {
            echo "  <td class='dele'>".
                '<a onclick="return confirm(\'Are you sure you want to delete device: '.$row['devicename'].'?\')" '.
                'href="'.$_SERVER['PHP_SELF'].'?action=delete&id='.$row['id'].'" class="dele">Delete</a>'.
                "</td>\n";
        }
        echo "<tr>\n";
    }
    
?>
</table>
<?php
}

function devices_header($subtitle = 'List Devices', $message = array())
{
?><!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html>
<head>
  <meta http-equiv="Content-Type" content="text/html; charset=iso-8859-1">
  <title>Manage Devices &gt; <?php echo $subtitle; ?></title>
<?php devices_css(); ?>
</head>
<body>
<h1>Manage Devices</h1>
<h2><?php echo $subtitle; ?></h2>
<?php
    foreach ($message as $s) :
?>
<strong><font color="red"><?php echo $s; ?></font></strong><br>
<?php
    endforeach;
}

function devices_footer()
{
?>
</body>
</html>
<?php
}

function devices_gen_pwd()
{
    /* Not including O, 0, 1, l to prevent ambiguity */
    $chars = "ABCDEFGHJKLMNPQRTYVWXYZabcdefghijkmnopqrstuvwxyz23456789";
    srand((double)microtime()*1000000);
    $i = 0;
    $pass = '' ;

    while ($i <= 17) {
        $num = rand() % strlen($chars);
        $tmp = substr($chars, $num, 1);
        $pass = $pass . $tmp;
        $i++;
    }

    return $pass;
}

function devices_ts_to_date($ts, $ft = '%a %b %e, %Y @ %r')
{
    global $db;

    $ts = mysql_real_escape_string($ts);
    $db->query("SELECT DATE_FORMAT('$ts','$ft')");
    $r = mysql_fetch_array($db->result);
    return $r[0];
}

function devices_css()
{
?>
  <style><!--
    a {
      color: #00f;
      text-decoration: none;
      text-align: center;
    }
    a:hover {
      background: #ccc;
      color: #00f;
    }
    body {
      background: #fff;
      color: #000;
      font: 10pt Georgia, "Times New Roman", Times, serif;
      padding-left: 30px;
      padding-right: 30px;
    }
    td.edit {
      text-align: center;
    }
    td.view {
      text-align: center;
    }
    td.dele {
      text-align: center;
    }
    th {
      background: #ddd;
      text-align: center;
    }
    tr.alt {
      background: #eee;
    }
    pre {
      margin-left: 10px;
      margin-right: 10px;
      background: #ddd;
    } -->
  </style>
<?php
}
?>
